# Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
import pandas as pd

data_path = 'data/supermarket_sales.csv'
df = pd.read_csv(data_path)

In [2]:
df

Invoice ID Branch       City Customer type  Gender  \
0    750-67-8428      A     Yangon        Member  Female   
1    226-31-3081      C  Naypyitaw        Normal  Female   
2    631-41-3108      A     Yangon        Normal    Male   
3    123-19-1176      A     Yangon        Member    Male   
4    373-73-7910      A     Yangon        Normal    Male   
..           ...    ...        ...           ...     ...   
995  233-67-5758      C  Naypyitaw        Normal    Male   
996  303-96-2227      B   Mandalay        Normal  Female   
997  727-02-1313      A     Yangon        Member    Male   
998  347-56-2442      A     Yangon        Normal    Male   
999  849-09-3807      A     Yangon        Member  Female   

               Product line  Unit price  Quantity   Tax 5%      Total  \
0         Health and beauty       74.69         7  26.1415   548.9715   
1    Electronic accessories       15.28         5   3.8200    80.2200   
2        Home and lifestyle       46.33         7  16.2155   340.5255   
3         Health and beauty       58.22         8  23.2880   489.0480   
4         Sports and travel       86.31         7  30.2085   634.3785   
..                      ...         ...       ...      ...        ...   
995       Health and beauty       40.35         1   2.0175    42.3675   
996      Home and lifestyle       97.38        10  48.6900  1022.4900   
997      Food and beverages       31.84         1   1.5920    33.4320   
998      Home and lifestyle       65.82         1   3.2910    69.1110   
999     Fashion accessories       88.34         7  30.9190   649.2990   

          Date   Time      Payment    cogs  gross margin percentage  \
0     1/5/2019  13:08      Ewallet  522.83                 4.761905   
1     3/8/2019  10:29         Cash   76.40                 4.761905   
2     3/3/2019  13:23  Credit card  324.31                 4.761905   
3    1/27/2019  20:33      Ewallet  465.76                 4.761905   
4     2/8/2019  10:37      Ewallet  604.17                 4.761905   
..         ...    ...          ...     ...                      ...   
995  1/29/2019  13:46      Ewallet   40.35                 4.761905   
996   3/2/2019  17:16      Ewallet  973.80                 4.761905   
997   2/9/2019  13:22         Cash   31.84                 4.761905   
998  2/22/2019  15:33         Cash   65.82                 4.761905   
999  2/18/2019  13:28         Cash  618.38                 4.761905   

     gross income  Rating  
0         26.1415     9.1  
1          3.8200     9.6  
2         16.2155     7.4  
3         23.2880     8.4  
4         30.2085     5.3  
..            ...     ...  
995        2.0175     6.2  
996       48.6900     4.4  
997        1.5920     7.7  
998        3.2910     4.1  
999       30.9190     6.6  

[1000 rows x 17 columns]

1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [3]:
profit_by_branch = df.groupby('Branch')['gross income'].agg([
    ('Total Profit', 'sum'),
    ('Average Profit per Sale', 'mean')
])

branch_sum = df.groupby('Branch')['gross income'].sum()
top_profit = branch_sum.idxmax()

avg_income = df.groupby('Branch')['gross income'].mean()
top_avg_income = avg_income.idxmax()

print("Profit Analysis by Branch:")
print(profit_by_branch)
print(f'\nBranch with Highest Profit: {top_profit}')

Profit Analysis by Branch:
        Total Profit  Average Profit per Sale
Branch                                       
A          5057.1605                14.874001
B          5057.0320                15.232024
C          5265.1765                16.052367

Branch with Highest Profit: C


2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

In [4]:
city_profit = df[df['Branch'] == top_profit]['City'].iloc[0]
print(f'City with Highest Profit: {city_profit}')

City with Highest Profit: Naypyitaw


The city of Naypyitaw has the highest total profit and the highest average receipt among all branches.

This indicates strong purchasing power of customers or a strategically successful store location.

I recommend considering the opening of an additional store in this city to scale the successful experience and further strengthen the company’s market presence in the region.

3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [5]:
invoices_by_category = df.pivot_table(index='Branch',columns ='Product line',values='Invoice ID', aggfunc='count')
invoices_by_category

Product line  Electronic accessories  Fashion accessories  Food and beverages  \
Branch                                                                          
A                                 60                   51                  58   
B                                 55                   62                  50   
C                                 55                   65                  66   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
Branch                                                                  
A                            47                  65                 59  
B                            53                  50                 62  
C                            52                  45                 45

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [6]:
def max_electronics_accessories(table,category):
  branch_max_invoice = table[category].idxmax()
  max_invoices = table[category].max()
  return (f"Branch with most Electronic accessories purchases: {branch_max_invoice}, number of purchases - {max_invoices}")

result = max_electronics_accessories(invoices_by_category,'Electronic accessories')
print(result)

Branch with most Electronic accessories purchases: A, number of purchases - 60


4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


**1. Customers with loyalty cards (Member) may generate more revenue for the company than regular customers (Normal)**

In [7]:
summary_table = df.groupby('Customer type').agg(
    Count = ('Invoice ID', 'count'),
    Average_Total = ('Total', 'mean'),
    Total_Revenue = ('Total', 'sum')
).round(2)

summary_table

Count  Average_Total  Total_Revenue
Customer type                                     
Member           501         327.79      164223.44
Normal           499         318.12      158743.30

**Insights based on the table:**
- Member customers:
    - Average check: 327.79
    - Total revenue: 164,223.44
- Normal customers:
    - Average check: 318.12
    - Total revenue: 158,743.30
- With nearly equal number of purchases (501 vs 499), Member clients:
    - spend more per purchase,
    - generate higher total revenue.

**Recommendation:**

Promote the loyalty program more actively.
- Offer sign-up bonuses.
- Encourage regular customers to become members.
- Analyze Member behavior to optimize future offers and upselling.

**2. Store profitability varies by day of the week, and some days may be significantly more profitable than others.**

In [8]:
# Step 1. Ensure Date is datetime
df['Date'] = pd.to_datetime(df['Date'])

# Step 2. Create a new column with weekday name
df['Day Of Week'] = df['Date'].dt.day_name()

# Step 3. Group by weekday and sum revenue
busiest_days = df.groupby('Day Of Week')['Total'].sum().reset_index().round(2)

# Step 4. Sort weekdays in correct order
day_order = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']
busiest_days['Day Of Week'] = pd.Categorical(busiest_days['Day Of Week'], categories=day_order, ordered=True)
busiest_days = busiest_days.sort_values('Day Of Week')

# Step 5. Show the result
busiest_days

Day Of Week     Total
1      Monday  37899.08
5     Tuesday  51482.25
6   Wednesday  43731.14
4    Thursday  45349.25
0      Friday  43926.34
2    Saturday  56120.81
3      Sunday  44457.89

**Insights based on the table:**

Saturday generates the highest revenue, while Monday is the least profitable. This may be due to weekend shopping habits, promotions, or customer behavior patterns.

**Recommendation:**
1. Intensify marketing on Fridays and Saturdays to boost traffic on peak days.
2. Introduce Monday-only promotions to improve sales on the slowest day.
3. Analyze Saturday customer behavior: identify which products/categories drive the most spending.


**3. Customers in different cities prefer different payment methods. These preferences may depend on demographics, technology adoption, or local consumer habits**

In [9]:
# Step 1. Group by City and Payment
analysis_payment_city = df.groupby(["City", "Payment"])["Total"].agg([
    ('Average Purchase', 'mean'),
    ('Count', 'count'),
    ('Total Revenue', 'sum')
]).round(2)

# Step 2. Calculate total count of transactions per city
total_counts_city = analysis_payment_city.groupby("City")["Count"].transform('sum')

# Step 3. Add percentage column (% of total transactions in each city)
analysis_payment_city["% of Total Count in City"] = (analysis_payment_city["Count"] / total_counts_city * 100).round(2)

# Step 4. Sort by City and percentage (descending)
analysis_payment_city = analysis_payment_city.sort_values(by=["City", "% of Total Count in City"], ascending=[True, False])

# Step 5. Show the result
analysis_payment_city

Average Purchase  Count  Total Revenue  \
City      Payment                                               
Mandalay  Ewallet                296.58    113       33513.35   
          Cash                   321.27    110       35339.46   
          Credit card            342.61    109       37344.86   
Naypyitaw Cash                   347.47    124       43085.86   
          Ewallet                350.52    106       37155.38   
          Credit card            309.46     98       30327.46   
Yangon    Ewallet                312.10    126       39324.37   
          Cash                   307.10    110       33781.25   
          Credit card            318.22    104       33094.75   

                       % of Total Count in City  
City      Payment                                
Mandalay  Ewallet                         34.04  
          Cash                            33.13  
          Credit card                     32.83  
Naypyitaw Cash                            37.80  
          Ewallet                         32.32  
          Credit card                     29.88  
Yangon    Ewallet                         37.06  
          Cash                            32.35  
          Credit card                     30.59

**Insights based on the table:**
1. In Naypyitaw, the most used payment method is Cash, representing 37.8% of all transactions in that city.
2. In Yangon, Ewallet is the most preferred method, accounting for 37.06% of transactions.
3. In Mandalay, all three payment methods are nearly equally distributed — each with around 33–34% share.

**Recommendations:**
1. Targeted promotions:
    - In Naypyitaw, consider launching discounts or cashback offers for digital payment methods (e.g. Ewallet or Credit card) to encourage their adoption.
2. POS infrastructure optimization:
    - In Yangon, it’s crucial to optimize systems for Ewallet payments, including fast QR scanning and support for mobile payments like Apple Pay or Google Pay.
3. Balanced approach in Mandalay:
    - Since all methods are similarly popular, ensure equal availability and support for all payment options.
4. Localized marketing strategy:
    - Tailor advertising messages based on local payment preferences, e.g., highlighting the convenience of Ewallet in Yangon.